source/key_derivation.c

```diff
@@ -0,0 +1,370 @@
+        return AWS_OP_ERR;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+/*
+ * aws_be_bytes_compare_constant_time() and aws_be_bytes_add_one_constant_time() are constant-time arithmetic functions
+ * that operate on raw bytes as if they were unbounded integers in a big-endian base 255 format.
+ */
+
+/*
+ * In the following function gt and eq are updated together.  After each update, the variables will be
+ * in one of the following states:
+ *
+ *  (1) gt is 0, eq is 1, and from an ordering perspective, lhs == rhs, as checked "so far"
+ *  (2) gt is 1, eq is 0, (lhs > rhs)
+ *  (3) gt is 0, eq is 0, (lhs < rhs)
+ *
+ *  States (2) and (3) are terminal states that cannot be exited since eq is 0 and is the and-wise mask of all
+ *  subsequent gt updates.  Similarly, once eq is zero it cannot ever become non-zero.
+ *
+ *  Intuitively these ideas match the standard way of comparing magnitude equality by considering digit count and
+ *  digits from most significant to least significant.
+ *
+ *  Let l and r be the the two digits that we are
+ *  comparing between lhs and rhs.  Assume 0 <= l, r <= 255 seated in 32-bit integers
+ *
+ *  gt is maintained by the following bit trick:
+ *
+ *      l > r <=>
+ *      (r - l) < 0 <=>
+ *      (r - l) as an int32 has the high bit set <=>
+ *      ((r - l) >> 31) & 0x01 == 1
+ *
+ *  eq is maintained by the following bit trick:
+ *
+ *      l == r <=>
+ *      l ^ r == 0 <=>
+ *      (l ^ r) - 1 == -1 <=>
+ *      (((l ^ r) - 1) >> 31) & 0x01 == 1
+ *
+ *  We apply to the volatile type modifier to attempt to prevent all early-out optimizations that a compiler might
+ *  apply if it performed constraint-based reasoning on the logic.  This is based on treating volatile
+ *  semantically as "this value can change underneath you at any time so you always have to re-read it and cannot
+ *  reason statically about program behavior when it reaches a certain value (like 0)"
+ */
+
+/**
+ * Compares two large unsigned integers in a raw byte format.
+ * The two operands *must* be the same size (simplifies the problem significantly).
+ *
+ * The output parameter comparison_result is set to:
+ *   -1 if lhs_raw_be_bigint < rhs_raw_be_bigint
+ *    0 if lhs_raw_be_bigint == rhs_raw_be_bigint
+ *    1 if lhs_raw_be_bigint > rhs_raw_be_bigint
+ */
+int aws_be_bytes_compare_constant_time(
+    const struct aws_byte_buf *lhs_raw_be_bigint,
+    const struct aws_byte_buf *rhs_raw_be_bigint,
+    int *comparison_result) {
+
+    AWS_FATAL_PRECONDITION(aws_byte_buf_is_valid(lhs_raw_be_bigint));
+    AWS_FATAL_PRECONDITION(aws_byte_buf_is_valid(rhs_raw_be_bigint));
+
+    /*
+     * We only need to support comparing byte sequences of the same length here
+     */
+    const size_t lhs_len = lhs_raw_be_bigint->len;
+    if (lhs_len != rhs_raw_be_bigint->len) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    volatile uint8_t gt = 0;
```

Comment:
Should probably just add a comment here explaining why these are volatile, it's not immediately obvious unless you really think about it

Reply:
line 137 explains, unsure if adding additional references would help.

Reply:
Oh good call, I missed that.